From 2b351290ec20b705106b68bf383105a0697ed87f Mon Sep 17 00:00:00 2001
From: Valentin Caron <valentin.caron@foss.st.com>
Date: Mon, 2 Sep 2024 15:28:59 +0200
Subject: [PATCH 1230/1695] plat-stm32mp2: scmi_server: implement dvfs protocol

scmi_server_scpfw can now retrieve opp description from DT.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 374bbb0a50616787d120442ae2d6044d28284a07 plat-stm32mp2: scmi_server: implement dvfs protocol]
Signed-off-by: Valentin Caron <valentin.caron@foss.st.com>
Change-Id: Ic879617ef425113a94899d580c316ea1f95d2d06
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/401864
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 .../arm/plat-stm32mp2/scmi_server_scpfw.c     |  6 ++
 core/drivers/stm32_cpu_opp.c                  | 71 ++++++++++++++-----
 core/include/drivers/stm32_cpu_opp.h          | 13 ++++
 .../include/scmi_agent_configuration.h        |  8 ---
 4 files changed, 73 insertions(+), 25 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c b/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
index de9e0c05e..43c67a1ab 100644
--- a/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
+++ b/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
@@ -6,6 +6,7 @@
 #include <drivers/mailbox.h>
 #include <drivers/scmi.h>
 #include <drivers/scmi-msg.h>
+#include <drivers/stm32_cpu_opp.h>
 #include <drivers/stm32_remoteproc.h>
 #include <drivers/stm32mp_dt_bindings.h>
 #include <kernel/boot.h>
@@ -405,6 +406,11 @@ static TEE_Result optee_scmi_server_probe(const void *fdt, int parent_node,
 		struct scpfw_channel_config *channel_cfg =
 			agent_cfg->channel_config + agent_ctx->channel_id;
 
+		res = optee_scmi_server_init_dvfs(fdt, 0, agent_cfg,
+						  channel_cfg);
+		if (res)
+			panic("Error during dvfs init");
+
 		SIMPLEQ_FOREACH(protocol_ctx, &agent_ctx->protocol_list, link) {
 			switch (protocol_ctx->protocol_id) {
 			case SCMI_PROTOCOL_ID_POWER_DOMAIN:
diff --git a/core/drivers/stm32_cpu_opp.c b/core/drivers/stm32_cpu_opp.c
index 311448af1..07723019f 100644
--- a/core/drivers/stm32_cpu_opp.c
+++ b/core/drivers/stm32_cpu_opp.c
@@ -379,15 +379,29 @@ static int max_cpu_voltage(struct cpu_dvfs *dvfs, size_t count)
 	return max_mv;
 }
 
-static void setup_scp_server_resources(void)
+TEE_Result optee_scmi_server_init_dvfs(const void *fdt __unused,
+				       int node __unused,
+				       struct scpfw_agent_config *agent_cfg,
+				       struct scpfw_channel_config *channel_cfg)
 {
 	TEE_Result res = TEE_ERROR_GENERIC;
 	unsigned int *dvfs_khz = NULL;
 	unsigned int *dvfs_mv = NULL;
+	unsigned int *dvfs_opp_khz = NULL;
+	unsigned int *dvfs_opp_mv = NULL;
 	size_t opp = 0;
 	struct cpu_dvfs *sorted_dvfs = NULL;
 	unsigned int scpfw_default_opp_index = UINT_MAX;
 
+	assert(agent_cfg && channel_cfg);
+
+	/*
+	 * Platform currenty expect only non-secure Cortex-A
+	 * (aka agent 1/channel 0) exposes a CPU DFVS service.
+	 */
+	if (agent_cfg->agent_id != 1 || channel_cfg->channel_id != 0)
+		return TEE_SUCCESS;
+
 	/*
 	 * Sort operating points by increasing frequencies as expected by
 	 * SCP-firmware DVFS module.
@@ -412,8 +426,10 @@ static void setup_scp_server_resources(void)
 	};
 
 	res = clk_register(&cpu_opp.scp_clock);
-	if (res)
-		panic();
+	if (res) {
+		free(sorted_dvfs);
+		return res;
+	}
 
 	/* Setup a regulator for scp-firmare DVFS module PSU instance */
 	cpu_opp.scp_regulator = (struct regulator){
@@ -453,22 +469,46 @@ static void setup_scp_server_resources(void)
 
 	free(sorted_dvfs);
 
-	res = scmi_scpfw_cfg_add_dvfs(0 /*agent*/, 0 /*channel*/, 0 /*domain*/,
-				      &cpu_opp.scp_regulator,
-				      &cpu_opp.scp_clock,
-				      dvfs_khz, dvfs_mv, cpu_opp.opp_count,
-				      scpfw_default_opp_index);
-	if (res)
-		panic();
+	/*
+	 * Fill struct scmi_perfd
+	 *
+	 * Only perfd[0] will be filled
+	 */
+	dvfs_opp_khz = calloc(cpu_opp.opp_count, sizeof(*dvfs_opp_khz));
+	dvfs_opp_mv = calloc(cpu_opp.opp_count, sizeof(*dvfs_opp_mv));
+
+	channel_cfg->perfd_count = 1;
+	channel_cfg->perfd = calloc(channel_cfg->perfd_count,
+				    sizeof(*channel_cfg->perfd));
+
+	if (!dvfs_opp_khz || !dvfs_opp_mv || !channel_cfg->perfd) {
+		free(dvfs_opp_mv);
+		free(dvfs_opp_khz);
+		free(cpu_opp.scp_cpu_opp_levels_uv);
+		free(channel_cfg->perfd);
+
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+	memcpy(dvfs_opp_mv, dvfs_mv, cpu_opp.opp_count * sizeof(*dvfs_opp_mv));
+	memcpy(dvfs_opp_khz, dvfs_khz,
+	       cpu_opp.opp_count * sizeof(*dvfs_opp_khz));
+
+	channel_cfg->perfd[0] = (struct scmi_perfd){
+		.name = "CPU DVFS",
+		.initial_opp = scpfw_default_opp_index,
+		.dvfs_opp_count = cpu_opp.opp_count,
+		.dvfs_opp_khz = dvfs_opp_khz,
+		.dvfs_opp_mv = dvfs_opp_mv,
+		.clk = &cpu_opp.scp_clock,
+		.regulator = &cpu_opp.scp_regulator,
+	};
 
 	free(dvfs_khz);
 	free(dvfs_mv);
+
+	return TEE_SUCCESS;
 }
 #else /*CFG_SCPFW_MOD_DVFS*/
-static void setup_scp_server_resources(void)
-{
-}
-
 TEE_Result stm32_cpu_opp_set_level(unsigned int level)
 {
 	unsigned int current_level = 0;
@@ -742,9 +782,6 @@ stm32_cpu_opp_init(const void *fdt, int cpu_node, int node,
 	if (res)
 		return res;
 
-	if (IS_ENABLED(CFG_SCPFW_MOD_DVFS))
-		setup_scp_server_resources();
-
 	return TEE_SUCCESS;
 }
 
diff --git a/core/include/drivers/stm32_cpu_opp.h b/core/include/drivers/stm32_cpu_opp.h
index 5a19e6d13..b220bc737 100644
--- a/core/include/drivers/stm32_cpu_opp.h
+++ b/core/include/drivers/stm32_cpu_opp.h
@@ -7,6 +7,9 @@
 #define DRIVERS_STM32_CPU_OPP_H
 
 #include <tee_api_types.h>
+#ifdef CFG_SCMI_SCPFW
+#include <scmi_agent_configuration.h>
+#endif /* CFG_SCMI_SCPFW */
 
 struct regulator;
 struct clk;
@@ -26,11 +29,21 @@ unsigned int stm32_cpu_opp_level(size_t opp_index);
 void scmi_server_set_cpu_resources(struct regulator *regulator, struct clk *clk,
 				   unsigned int *dvfs_khz,
 				   unsigned int *dvfs_mv, size_t dvfs_count);
+
+/*
+ * Initialize SCMI DVFS from DT.
+ * Returns a TEE_Result compliant value
+ */
+TEE_Result optee_scmi_server_init_dvfs(const void *fdt, int node,
+				       struct scpfw_agent_config *agent_cfg,
+				       struct scpfw_channel_config *channel_cfg
+				       );
 #else
 /* Request to switch to CPU operating point related to @level */
 TEE_Result stm32_cpu_opp_set_level(unsigned int level);
 
 /* Get level related to current CPU operating point */
 TEE_Result stm32_cpu_opp_read_level(unsigned int *level);
+
 #endif /*CFG_SCPFW_MOD_DVFS*/
 #endif /*DRIVERS_STM32_CPU_OPP_H*/
diff --git a/core/lib/scmi-server/include/scmi_agent_configuration.h b/core/lib/scmi-server/include/scmi_agent_configuration.h
index 72989de3d..382c6416f 100644
--- a/core/lib/scmi-server/include/scmi_agent_configuration.h
+++ b/core/lib/scmi-server/include/scmi_agent_configuration.h
@@ -163,14 +163,6 @@ struct scpfw_config *scmi_scpfw_get_configuration(void);
 /* Release resources allocated to create SCP-firmware configuration data */
 void scmi_scpfw_release_configuration(void);
 
-/* Add a DVFS service in the SCMI server */
-TEE_Result scmi_scpfw_cfg_add_dvfs(unsigned int agent, unsigned int channel,
-				   unsigned int domain_id,
-				   struct regulator *regulator, struct clk *clk,
-				   unsigned int *dvfs_khz,
-				   unsigned int *dvfs_mv, size_t dvfs_count,
-				   size_t initial_index);
-
 /* SCP firmware SCMI server configuration entry point */
 void scpfw_configure(struct scpfw_config *cfg);
 
-- 
2.47.2

